File: src/enemy.cpp
#include "enemy.hpp"
#include "settings.hpp"
#include <experimental/filesystem>
#include <algorithm>

std::pair<float, py::Vector2f> get_object_distance_direction(
    py::Rect<float> object,
    py::Rect<float> target)
{
    auto object_vec = py::Vector2f(object.center());
    auto target_vec = py::Vector2f(target.center());
    auto distance = object_vec.distance_to(target_vec);
    auto direction = (distance > 0) ? py::Vector2f(target_vec - object_vec).normalize() : py::Vector2f(0, 0);
    return {distance, direction};
}

Enemy::Enemy(const std::string &monster_name,
             const sf::Vector2f &pos,
             const std::list<std::shared_ptr<SpriteTexture>> &obstacle_sprites,
             const std::function<void(const sf::Vector2f, const std::string &)> &trigger_death_particles,
             const std::function<void(int)> &add_exp,
             SpriteManager &sprite_manager,
             std::shared_ptr<Player> player)
    : Entity(0, 0.15, {0, 0}, obstacle_sprites),
      monster_name(monster_name),
      can_attack(true),
      attack_cooldown(sf::milliseconds(400)),
      vulernable(true),
      invincibility_duration(sf::milliseconds(300)),
      trigger_death_particles(trigger_death_particles),
      add_exp(add_exp),
      sprite_manager(sprite_manager),
      player(player)
{
    this->sprite_type_ = "enemy";
    this->import_graphics(monster_name);
    this->status = std::make_shared<EnemyIdleState>(this->direction);

    auto name = monster_name + "/" + this->status->get();
    auto &animation = this->sprite_manager.textures(name);
    this->update_sprite(animation[static_cast<int>(this->frame_index)]);
    // set object rectangle
    auto rect = this->sprite_.getTextureRect();
    this->rect_ = py::Rect<float>(rect.left, rect.top, rect.width, rect.height);
    this->get_rect({"topleft", pos});
    this->hitbox_ = this->rect_.inflate(0, -26);

    auto monster_info = std::find_if(MONSTER_DATA.begin(), MONSTER_DATA.end(),
                                     [&monster_name](const std::pair<std::string, MonsterData> &m)
                                     { return !m.first.compare(monster_name); });
    this->health = monster_info->second.health;
    this->exp = monster_info->second.exp;
    this->damage = monster_info->second.damage;
    this->attack_type = monster_info->second.attack_type;
    this->speed = monster_info->second.speed;
    this->resistance = monster_info->second.resistance;
    this->attack_radius = monster_info->second.attack_radius;
    this->notice_radius = monster_info->second.notice_radius;

    this->death_sound_buffer.loadFromFile("./src/audio/hit.wav");
    this->death_sound.setBuffer(this->death_sound_buffer);
    this->death_sound.setVolume(10);

    this->hit_sound_buffer.loadFromFile("./src/audio/hit.wav");
    this->hit_sound.setBuffer(this->hit_sound_buffer);
    this->hit_sound.setVolume(10);

    this->attack_sound_buffer.loadFromFile(monster_info->second.attack_sound);
    this->attack_sound.setBuffer(attack_sound_buffer);
    this->attack_sound.setVolume(10);
}

void Enemy::import_graphics(const std::string &monster_name)
{
    auto base_path = std::experimental::filesystem::path("./src/graphics/monsters");
    sprite_manager.import(this->monster_name + "/idle", base_path / this->monster_name / "idle");
    sprite_manager.import(this->monster_name + "/move", base_path / this->monster_name / "move");
    sprite_manager.import(this->monster_name + "/attack", base_path / this->monster_name / "attack");
}

void Enemy::get_status(std::shared_ptr<Player> player)
{
    auto distance = get_object_distance_direction(this->rect_, player->rect()).first;
    if (this->can_attack && distance < this->attack_radius)
    {
        if (!this->status->eq("attack"))
        {
            this->frame_index = 0;
            this->status = std::make_shared<EnemyAttackState>(
                this->player,
                this->attack_time,
                this->attack_sound,
                this->damage,
                this->attack_type);
        }
    }
    else if (distance < this->notice_radius)
    {
        this->status = std::make_shared<EnemyMoveState>(
            this->direction,
            this->rect_,
            this->player);
    }
    else
    {
        if (!this->status->eq("idle"))
        {
            this->status = std::make_shared<EnemyIdleState>(
                this->direction);
        }
    }
}

void Enemy::actions(std::shared_ptr<Player> player)
{
    this->status->action();
}

void Enemy::animate()
{
    auto name = monster_name + "/" + this->status->get();
    auto &animation = this->sprite_manager.textures(name);

    this->frame_index += this->animation_speed;
    if (this->frame_index >= static_cast<float>(animation.size()))
    {
        if (this->status->eq("attack"))
        {
            this->can_attack = false;
        }
        this->frame_index = 0;
    }

    this->update_sprite(animation[static_cast<int>(this->frame_index)]);
    this->get_rect({"center", this->hitbox_.center()});

    if (!this->vulernable)
    { // make flicker
        auto alpha = this->wave_value();
        this->sprite_.setColor(sf::Color(0xff, 0xff, 0xff, alpha));
    }
    else
    {
        this->sprite_.setColor(sf::Color(0xff, 0xff, 0xff, 0xff));
    }
}

void Enemy::update()
{
    this->hit_reaction();
    this->move(this->speed);
    this->animate();
    this->cooldowns();
    this->check_death();

    this->get_status(this->player);
    this->actions(this->player);
}

void Enemy::get_damage(std::shared_ptr<Player> player, const std::string &attack_type)
{
    if (this->vulernable)
    {
        this->hit_sound.play();
        this->direction = get_object_distance_direction(this->rect_, player->rect()).second;
        if (!attack_type.compare("weapon"))
        {
            this->health -= player->get_full_weapon_damage();
        }
        else
        {
            this->health -= player->get_full_magic_damage();
        }
        this->hit_time.restart();
        this->vulernable = false;
    }
}

void Enemy::hit_reaction()
{
    if (!this->vulernable)
    {
        this->direction.x *= (this->resistance * -1);
        this->direction.y *= (this->resistance * -1);
    }
}

void Enemy::check_death()
{
    if (this->health <= 0)
    {
        this->death_sound.play();
        this->trigger_death_particles(this->rect_.center(), this->monster_name);
        this->sprite_manager.kill(this); // remove the Sprite from all Groups
        this->add_exp(this->exp);
    }
}

void Enemy::cooldowns()
{
    if (!this->can_attack)
    {
        if (this->attack_time.getElapsedTime() > this->attack_cooldown)
        {
            this->can_attack = true;
        }
    }

    if (!this->vulernable)
    {
        if (this->hit_time.getElapsedTime() > this->invincibility_duration)
        {
            this->vulernable = true;
        }
    }
}

void Enemy::update_sprite(const sf::Texture &texture)
{
    this->sprite_.setTexture(texture);
}

File: pyZelda/enemy.py
import pygame

from entity import Entity
from support import import_folder
from settings import monster_data
from player import Player

class Enemy(Entity):
    def __init__(self, monster_name, pos, groups, obstacles_sprites, trigger_death_particles, add_exp) -> None:
        super().__init__(groups)

        self.sprite_type = 'enemy'
        self.trigger_death_particles = trigger_death_particles
        self.add_exp = add_exp

        # graphics
        self.import_graphics(monster_name)
        self.status = 'idle'
        self.image = self.animations[self.status][int(self.frame_index)]

        # movement
        self.rect = self.image.get_rect(topleft=pos)
        self.hitbox = self.rect.inflate(0, -10)
        self.obstacles_sprites = obstacles_sprites

        # stats
        self.monster_name = monster_name
        monster_info = monster_data[self.monster_name]
    
        self.health = monster_info['health'] 
        self.exp = monster_info['exp']
        self.damage = monster_info['damage'] 
        self.attack_type = monster_info['attack_type'] 
        self.speed = monster_info['speed'] 
        self.resistance = monster_info['resistance'] 
        self.attack_radius = monster_info['attack_radius'] 
        self.notice_radius = monster_info['notice_radius'] 

        # player interaction
        self.can_attack = True
        self.attack_time = None
        self.attack_cooldown = 400

        # invincibility timer
        self.vulernable = True
        self.hit_time = None
        self.invincibility_duration = 300

        # sounds
        self.death_sound = pygame.mixer.Sound(f"./audio/hit.wav")
        self.hit_sound = pygame.mixer.Sound(f"./audio/hit.wav")
        self.attack_sound = pygame.mixer.Sound(monster_info['attack_sound'])
        self.death_sound.set_volume(0.1)
        self.hit_sound.set_volume(0.1)
        self.attack_sound.set_volume(0.1)

    def import_graphics(self, monster_name):
        self.animations = {
            "idle":[],
            "move":[],
            "attack": [],
        }

        for animation in self.animations.keys():
            self.animations[animation] = import_folder(f"./graphics/monsters/{monster_name}/{animation}")

    def get_player_distance_direction(self, player: Player):
        enemy_vec = pygame.math.Vector2(self.rect.center)
        player_vec = pygame.math.Vector2(player.rect.center)
        distance = enemy_vec.distance_to(player_vec)
        if distance > 0:
            direction = (player_vec-enemy_vec).normalize()
        else:
            direction = pygame.math.Vector2(0,0)

        return (distance, direction)

    def get_status(self, player):
        distance = self.get_player_distance_direction(player)[0]

        if distance < self.attack_radius and self.can_attack:
            if self.status != 'attack':
                self.frame_index = 0
            self.status = 'attack'
        elif distance < self.notice_radius:
            self.status = 'move'
        else:
            self.status = 'idle'

    def actions(self, player):
        if self.status == 'attack':
            self.attack_sound.play()
            self.attack_time = pygame.time.get_ticks()
            player.damage_player(self.damage, self.attack_type)
        elif self.status == 'move':
            self.direction = self.get_player_distance_direction(player)[1]
        else:
            self.direction = pygame.math.Vector2(0,0)

    def cooldowns(self):
        current_time = pygame.time.get_ticks()

        if not self.can_attack:
            if current_time - self.attack_time > self.attack_cooldown:
                self.can_attack = True

        if not self.vulernable:
            if current_time - self.hit_time >= self.invincibility_duration:
                self.vulernable = True

    def hit_reaction(self):
        if not self.vulernable:
            self.direction *= -self.resistance

    def get_damage(self, player, attack_type):
        if self.vulernable:
            self.hit_sound.play()
            self.direction = self.get_player_distance_direction(player)[1]
            if attack_type == 'weapon':
                self.health -= player.get_full_weapon_damage()
            else:
                self.health -= player.get_full_magic_damage()
            self.hit_time = pygame.time.get_ticks()
            self.vulernable = False

    def check_death(self):
        if self.health <= 0:
            self.death_sound.play()
            self.trigger_death_particles(self.rect.center, self.monster_name)
            self.kill()
            self.add_exp(self.exp)

    def animate(self):
        animation = self.animations[self.status]

        self.frame_index += self.animation_speed
        if self.frame_index >= len(animation):
            if self.status == 'attack':
                self.can_attack = False
            self.frame_index = 0

        self.image = animation[int(self.frame_index)]
        self.rect = self.image.get_rect(center=self.hitbox.center)

        if not self.vulernable:
            # flicker
            alpha = self.wave_value()
            self.image.set_alpha(alpha)
        else:
            self.image.set_alpha(255)


    def update(self):
        self.hit_reaction()
        self.move(self.speed)
        self.animate()
        self.cooldowns()
        self.check_death()

    def enemy_update(self, player: Player):
        self.get_status(player)
        self.actions(player)